### ISSUE TITLE
Version specification limits

### ISSUE THREAD

**Issue by bmewing** [bmewing/requirements#59]
I discovered today while writing some tests that you can't specify versions like `mgsub > 2`, you have to specify them like `mgsub > 2.0` or you'll get an `invalid version specification` error from `package_version`

**Comment by bmewing** [comment]
The question is, where do we want to handle this? Modify invalid version numbers to include a `.0`? Throw an error when parsing the file? Something else?

**Comment by AdamSpannbauer** [comment]
I wrote the function `validate_version()` to handle this but not exactly sure where it needs to fit in to the current code.  See the tests below to understand the curent behavior I chose for this first pass.

```r
library(testthat)

test_that("validate_version", {
  expect_equal(validate_version("2"), package_version("2.0"))
  expect_equal(validate_version("2.0"), package_version("2.0"))

  expect_error(validate_version("x"), regexp = "invalid version specification")
})

```
```r
safe_version = function(x) {
  #' @keywords internal
  #' Gracefully fail coercing strings with \code{package_version()}
  #'
  #' @param x string to coerce to \code{package_version}
  #' @return \code{package_version} object if successful; NULL if unsuccessful

  tryCatch(
    expr = {
      return(package_version(x, strict = TRUE))
    },
    error = function(e) {
      return(NULL)
    }
  )
}


validate_version = function(x) {
  #' @keywords internal
  #' Coerce strings to \code{package_version} with \code{package_version()}
  #'
  #' @details This function's purpose is to validate integer version strings
  #'
  #' @param x string to coerce to \code{package_version}
  #' @return \code{package_version} object

  validated_verion = safe_version(x)
  if (!is.null(validated_verion)) return(validated_verion)

  validated_verion = safe_version(paste0(x, ".0"))
  if (!is.null(validated_verion)) return(validated_verion)

  stop(sprintf("invalid version specification ‘%s’", x))
}

```

---

EDIT: deleted the `int_to_version` helper. seemed overkill to break it out on its own. should be in comment edit history if we want it back

**Comment by bmewing** [comment]
the code looks good. like you say, we just need to figure out where in the flow to put it.

**Comment by AdamSpannbauer** [comment]
Documenting here that I'm thinking the current helper function `safe_package_version()` should be renamed to `safe_packageVersion()` and the `safe_version()` function above should be named `safe_package_version()`.

I think this is the clearest way to know which base functions they relate to since it will be the base name with `safe_` prepended.  We'll just have to tell lintr-bot to eff off on the usage of camelcase.

**Comment by bmewing** [comment]
#63 has a function to be used in `process_requirements` to flag malformed requirements and throw an error.

**Comment by AdamSpannbauer** [comment]
To clarify based on what's in #63 

We prefer to request the user to change the version requirement specification of `mgsub > 2` instead of assuming that this is equivalent to `mgsub > 2.0`?

If this is the case I feel this might be inconsistent with auto-correcting the `=` to `==`.  My vote would be to either auto-correct both or to fail both.

**Comment by AdamSpannbauer** [comment]
To follow up on the above with the behavior of `pip install -r`.

#### `requirements.txt` contents
```
setuptools > 1
vidstab = 1.5.6
```

#### `pip install -r requirements.txt` output
```
Invalid requirement: 'vidstab = 1.5.6'
= is not a valid operator. Did you mean == ?
```

(the `setuptools > 1` is honored when i remove the invalid `vidstab` requirement)

---

Edit: so I change my vote to be consistent with `pip`.

**Comment by bmewing** [comment]
Has this been fully addressed by #63 or is something left to do @AdamSpannbauer?
